prov/verbs, rxm: avoid CQ overrun and empty receive queue

a-ilango opened the pull request:
The Infiniband spec doesn't say if the send queue space should be freed after
a WR is processed. Based on the implementation, we may need to track available
space in CQ to avoid overflowing it (which renders any QPs bound to it unusable).
Use an atomic counter to track available space and return -FI_EAGAIN on
ibv_post_send when we don't have credits.


@mmarcini, this PR updates libfabric verbs provider to avoid CQ overrun.

shefty commented on prov/verbs/src/fi_verbs.h:
```diff
@@ -354,6 +362,8 @@ struct fi_ibv_domain {
 	fi_ibv_mr_dereg_cb		internal_mr_dereg;
 	struct fi_ibv_mem_notifier	*notifier;
+	fi_ibv_post_send_func		post_send;
+	fi_ibv_poll_cq_func		poll_cq;
```
You don't need to create typedefs.  You can just define the function pointers as part of struct fi_ibv_domain  This is probably true for the mr callbacks as well.

Also, you may be able to set post_send=ibv_post_send(), and poll_cq=ibv_poll_cq() and avoid the wrappers when calling verbs directly.  To do this you would need pass in the ibv_qp and ibv_cq directly to the calls, and be able to use the ibv_qp->context and ibv_cq->context to get back to your structures for the tracking versions.  Hopefully that makes sense.

In fact, if you go further, you might be able to set the function pointers to call the ibv verbs provider post/poll calls directly and avoid the ibv_post/poll wrappers.

a-ilango replied:
>You don't need to create typedefs. You can just define the function pointers as part of struct fi_ibv_domain This is probably true for the mr callbacks as well.
Also, you may be able to set post_send=ibv_post_send(), and poll_cq=ibv_poll_cq() and avoid the wrappers when calling verbs directly. To do this you would need pass in the ibv_qp and ibv_cq directly to the calls, and be able to use the ibv_qp->context and ibv_cq->context to get back to your structures for the tracking versions. Hopefully that makes sense.

👍 
>In fact, if you go further, you might be able to set the function pointers to call the ibv verbs provider post/poll calls directly and avoid the ibv_post/poll wrappers.

We would still need the wrappers that check / update the credits, right?

shefty replied:
Yes, the wrappers for the credits are still needed.  Those would basically look like what you have already.

shefty commented on prov/verbs/src/fi_verbs.h:
```diff
@@ -944,23 +964,74 @@ static inline int fi_ibv_poll_reap_unsig_cq(struct fi_ibv_ep *ep)
+				   struct ibv_send_wr **bad_wr)
+{
+	return ibv_post_send(ep->ibv_qp, wr, bad_wr);
+}
```
nit: the static inlines for any of the handlers won't help because these are all accessed via function pointers.

shefty commented on prov/verbs/src/fi_verbs.h:
```diff
@@ -944,23 +964,74 @@ static inline int fi_ibv_poll_reap_unsig_cq(struct fi_ibv_ep *ep)
+{
+	int ret = ibv_poll_cq(cq->cq, num_entries, wc);
+	if (ret > 0) {
+		switch (wc->opcode) {
```
I think you can use if (!(wc->opcode & IBV_WC_RECV)) in place of a switch statement.

a-ilango replied:
Okay, we would avoid IBV_WC_RECV and IBV_WC_RECV_RDMA_WITH_IMM as well, assuming some new recv status won't be added in the future.

shefty replied:
The API is designed with the above check in mind.  So new recv opcodes would likely still work.

a-ilango commented:
@shefty I've updated the PR based on requested changes. Also added a patch to re-post receives to MSG provider ASAP to avoid hangs in rendezvous protocol.

As we discussed earlier, in-order to limit # of outstanding RNDV sends (to keep memory usage low), we need to track peer RX queue length. I think we can exchange this info as part of CM data during connection establishment. I'll add that in a follow-up PR. Do we need this CM data change in v1.7.x? Is it okay to update a protocol in stable branch?

shefty commented:
I would like to avoid protocol changes in the stable branch, unless we have no alternative.  Could we use some reasonable hard-coded value for 1.7?

shefty commented on prov/verbs/src/verbs_domain.c:
```diff
@@ -385,6 +385,44 @@ static void fi_ibv_domain_process_exp(struct fi_ibv_domain *domain)
+		FI_DBG(&fi_ibv_prov, FI_LOG_EP_DATA, "CQ credits not available,"
+		       " retry later\n");
+		ofi_atomic_inc32(&cq->credits);
+		return ENOMEM;
```
Is this the right return value?  Should this be negative or set errno instead?

a-ilango replied:
This tries to mimic ibv_post_send which returns ENOMEM on full send queue. `fi_ibv_handle_post` should convert a positive error value to appropriate libfabric error number.

shefty commented on prov/verbs/src/verbs_domain.c:
```diff
@@ -494,6 +532,14 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
 		goto err3;
 	}
 
+	if (!strncmp(info->domain_attr->name, "hfi1", strlen("hfi1"))) {
```
Do we need to add checks for the previous generations (before OPA)?  I don't recall what those were called though.

a-ilango replied:
I think it was `qib` but I'm not sure what's the behavior of its verbs implementation.

mmarcini replied:
In terms of completions qib shares the same rdmavt completion logic.

The issue should be there as well.

a-ilango replied:
> In terms of completions qib shares the same rdmavt completion logic.
The issue should be there as well.

I'll do a follow-up patch add qib.

a-ilango commented:
> I would like to avoid protocol changes in the stable branch, unless we have no alternative. Could we use some reasonable hard-coded value for 1.7?

We can probably hard code it to half the RX queue length assuming no shared context is used which is the default.
